src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,73 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, iterprod, idx)
+    for v in iterprod
+        push!(col, v[idx])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols; error=true::Bool)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    if !(indexcols isa Symbol || allunique(indexcols))
+        throw(ArgumentError("Elements of $indexcols must be unique"))
+    end
+    if error == true
+        row_group_slots(ntuple(i -> df[!, i], ncol(df)))[1]!=nrow(df) && throw(ArgumentError("duplicate rows in input"))
+    else
+        row_group_slots(ntuple(i -> df[!, i], ncol(df)))[1]!=nrow(df) && @warn "duplicate rows in input; expand will only return unique combinations"
```

Comment:
@nalimilan - can you please double check that `row_group_slots` is used correctly, as you know this part of code better.

Reply:
Sounds correct. As I noted above, passing `Val(false)` as the second argument can make thing faster in some cases, so better do that.

Also, I'd make the `row_group_slot` call the top-level branch, and put `if error` inside it: that would avoid repeating the call. And while I'm at it, some spacing around `!=` wouldn't hurt.